Memory allocation

awni opened the pull request:
## Proposed changes

This changes a few problems I observed in our memory allocator.

1. In some cases the cache is really full but nothing fits. Instead of trying to clear the cache, the allocator waits and in some extreme cases even fails when there is memory technically "available" but it's held by the cache. (If there are no scheduled jobs, nothing in the cache fits, we don't clear the cache, we just fail).

2. Change the cache fit policy to be a little stricter. It used to require the block to be less than twice the size. Now it requires `min(2 * size, size + vm_page_size)`. The `vm_page_size` is somewhat arbitrary but I think it's much better policy. Can save a lot of memory and when you get to big sizes, typically missing allocator the cache isn't such a huge deal as we fall back to allocating a new bugger.

3. Add a force swap parameter to the `malloc`. Once there are no more jobs to wait for, instead of just breaking the allocator will swap to the CPU. This makes it possible to do large model conversions for example.

4. Decrease the wait limit. I notice we often swap even at `1.5 * recommendedSize` (probably related to the amount of wired memory and other apps running). I cranked it down to just `recommendedSize`. That + the force allocate seems to work just as well in the cases I tested.

I will add some benchmarks before we decide to merge this. But want to solicit comments in parallel.
 
## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

awni commented:
### Before change

**Llama benchmark**
Time per layer per token: 3.9148330688476562 ms
Lower bound total time per token: 125.274658203125 ms

**Linear regression**
Loss 0.00005, |w-w*| = 0.00364, Throughput 3666.22979 (it/s)
Logistic regression
Loss 0.00005, |w-w*| = 0.00364, Throughput 3655.93535 (it/s)

**Mnist CPU**
Epoch 9: Test accuracy 0.948, Time 0.078 (s)

**Mnist GPU**
Epoch 9: Test accuracy 0.948, Time 0.168 (s)

**Transformer LM Batch Size 2**
Iter 20: Train loss 8.854, It/sec 2.291

**Transformer LM Batch Size 4**
Iter 20: Train loss 8.869, It/sec 1.198

**Transformer LM Batch Size 8**
Iter 20: Train loss 8.885, It/sec 0.599

### After change

**Llama benchmark**
Time per layer per token: 3.997039794921875 ms
Lower bound total time per token: 127.9052734375 ms

**Linear regression**
Loss 0.00005, |w-w*| = 0.00364, Throughput 3708.16439 (it/s)

**Logistic regression**
Loss 0.00005, |w-w*| = 0.00364, Throughput 3691.89505 (it/s)

**Mnist CPU**
Epoch 9: Test accuracy 0.948, Time 0.081 (s)

**Mnist GPU**
Epoch 9: Test accuracy 0.948, Time 0.167 (s)

**Transformer LM Batch Size 2**
Iter 20: Train loss 8.885, It/sec 2.290

**Transformer LM Batch Size 4**
Iter 20: Train loss 8.830, It/sec 1.194

**Transformer LM Batch Size 8**
Iter 20: Train loss 8.870, It/sec 0.530

angeloskath reviewed (commented):
This looks great, but I do get a pretty big performance regression on memory pressured transformer training on my M2 air.

awni commented:
Can you say more? What parameter(s) did you increase to get that behavior?

angeloskath commented:
Just the default :-) . On my M2 air it goes from a consistent 0.73 it/s for at least 60 iterations to ~0.63 it/s but it is also erratic in the sense that sometimes it starts low and slowly increases after several iterations or it starts ok at 0.72 and decreases. It has to do with the caching and waiting patterns...

When reducing the memory pressure by reducing the number of blocks it seems to perform the same. On the stable diffusion example it also seems to perform practically the same.

awni commented:
> It has to do with the caching and waiting patterns...

Yes almost certainly. I changed a couple things there. In general it's going to be hard to find a setting that makes everything faster :\, but I will see if I can tune it a bit for the transformer case.

awni commented:
@angeloskath I fixed one bug I introduced and partially reverted some of the GC strategy. I don't see an regression running the transformer benchmark with the default params on an 8GB machine (which is pretty high memory presssure). Could you pull and try running your benchmark again?

angeloskath reviewed (approved):
I think it is fine to merge!

We still have to improve our whole memory management cause it is quite sensitive (as you said) on the parameters of the allocator and the memory access patterns. For instance changing the block limit to a smaller number doesn't affect the transformer but it does affect the stable diffusion quite a lot. Eg making the decoder 5x slower. 

The current parameters seem to work very similarly to before so it should be fine to merge.
